polynote-server/src/main/scala/polynote/server/repository/fs/DatabaseFilesystem.scala

```diff
@@ -0,0 +1,443 @@
+        val pathAndName = extractPathAndName(removeExtension(path))
+        database.existsNotebookInDatabase(pathAndName._1, pathAndName._2 )
+    }
+
+    /* This is NOT an endpoint in our superclass, so we don't need a version of it here.
+        We certainly don't want to be creating any directories.  Stub it out.
+    */
+    private def createDirs(path: Path): RIO[BaseEnv, Unit] = effectBlocking( () )
+        /* Was:  effectBlocking(Files.createDirectories(path.getParent)) */
+
+    /** pair of functions that implement Rename/Move */
+    override def move(from: Path, to: Path): RIO[BaseEnv, Unit] =
+        effectBlocking( moveNotebookInDatabase(from.toFile.getPath, to.toFile.getPath) ).uninterruptible
+               /* was:  createDirs(to).map(_ => Files.move(from, to)) */
+
+    private def moveNotebookInDatabase(pathFrom: String, pathTo: String): Unit = {
+        val pathAndNameFrom = extractPathAndName(removeExtension(pathFrom))
+        val pathAndNameTo = extractPathAndName(removeExtension(pathTo))
+        database.moveNotebookInDatabase(pathAndNameFrom._1, pathAndNameFrom._2, pathAndNameTo._1, pathAndNameTo._2 )
+    }
+
+    /* pair of functions that implement Copy */
+    override def copy(from: Path, to: Path): RIO[BaseEnv, Unit] =
+        effectBlocking( copyNotebookInDatabase(from.toFile.getPath, to.toFile.getPath) ).uninterruptible
+        /* was:  createDirs(to).map(_ => Files.copy(from, to))  */
+
+    private def copyNotebookInDatabase(pathFrom: String, pathTo: String): Unit = {
+        val pathAndNameFrom = extractPathAndName(removeExtension(pathFrom))
+        val pathAndNameTo = extractPathAndName(removeExtension(pathTo))
+        database.copyNotebookInDatabase(pathAndNameFrom._1, pathAndNameFrom._2, pathAndNameTo._1, pathAndNameTo._2 )
+    }
+
+    /* pair of functions that implement Delete */
+    override def delete(path: Path): RIO[BaseEnv, Unit] =
+        exists(path).flatMap {
+          case false => ZIO.fail(new FileNotFoundException(s"File $path doesn't exist"))
+          case true  => effectBlocking( deleteNotebookInDatabase( path.toFile.getPath ) ).uninterruptible
+                         /* was:  effectBlocking(Files.delete(path)) */
+        }
+
+    private def deleteNotebookInDatabase(path: String ): Unit = {
+        val pathAndName = extractPathAndName(removeExtension(path))
+        database.deleteNotebookInDatabase(pathAndName._2, pathAndName._1 )
+    }
+
+    override def init(path: Path): RIO[BaseEnv, Unit] = {
+        dbContext = databaseContext.init()
+        database = new Database(dbContext)
+        createDirs(path)
+    }
+    override def createLog(path: Path): RIO[BaseEnv, WAL.WALWriter] =
+        effectBlocking(path.getParent.toFile.mkdirs()) *> FileChannelWALWriter(path)
+
+}
+
+/**
+ * Localized database access functions.  Here is where the actual database calls are made.
+ */
+class Database(dbContext: DbContext) {
```

Comment:
If you suspend all of the database effects here in this class, you can ensure they're all captured in the filesystem.